datastore/datastore.go

```diff
@@ -0,0 +1,303 @@
+	Sdc StackDriverImpl
+}
+
+type DatastoreImpl interface {
+	GetAll(ctx context.Context, q *datastore.Query, dst interface{}) (keys []*datastore.Key, err error)
+}
+
+type StackDriverImpl interface {
+	List(name string) *monitoring.ProjectsTimeSeriesListCall
+}
+
+type KeyData struct {
+	AuthProviderX509CertURL string  `json:"auth_provider_x509_cert_url"`
+	AuthURI                 string  `json:"auth_uri"`
+	ClientEmail             string  `json:"client_email"`
+	ClientID                float64 `json:"client_id,string"`
+	ClientX509CertURL       string  `json:"client_x509_cert_url"`
+	PrivateKey              string  `json:"private_key"`
+	PrivateKeyID            string  `json:"private_key_id"`
+	ProjectID               string  `json:"project_id"`
+	TokenURI                string  `json:"token_uri"`
+	Type                    string  `json:"type"`
+}
+
+// InventoryData is the data type for inventory data produced by a plugin data
+// source and emitted to the agent's inventory data store
+type InventoryData map[string]interface{}
+
+// MetricData is the data type for events produced by a plugin data source and
+// emitted to the agent's metrics data store
+type MetricData map[string]interface{}
+
+// EventData is the data type for single shot events
+type EventData map[string]interface{}
+
+// NAME - name of plugin
+const NAME string = "datastore"
+
+// ProtocolVersion -
+const ProtocolVersion string = "1"
+
+func Run(log *logrus.Logger, prettyPrint bool, version string) {
+	var data = PluginData{
+		Name:            NAME,
+		ProtocolVersion: ProtocolVersion,
+		PluginVersion:   version,
+		Inventory:       make(map[string]InventoryData),
+		Metrics:         make([]MetricData, 0),
+		Events:          make([]EventData, 0),
+	}
+
+	var keyData KeyData
+
+	keyFile, err := ioutil.ReadFile("/var/secrets/google/key.json")
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = json.Unmarshal(keyFile, &keyData)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", "/var/secrets/google/key.json")
+
+	projectID := keyData.ProjectID
+	c := NewClient(projectID)
+
+	//add stackdriver metrics
+	for _, metric := range stackdriverEndpoints {
+		resp, err := getStackdriverResp(c.Sdc, projectID, metric)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		result, err := getStackdriverData(resp)
+		if err != nil {
+			log.Fatal(err)
+		}
+		for _, metricResult := range result {
+			data.Metrics = append(data.Metrics, metricResult)
+		}
+	}
+
+	//add query metrics
+	kinds, err := getDatastoreQueryResult(c.Dsc)
+	if err != nil {
+		log.Fatal(err)
+	}
+	result := getDatastoreData(kinds, projectID)
+
+	for _, metricResult := range result {
+		data.Metrics = append(data.Metrics, metricResult)
+	}
+
+	fatalIfErr(log, helpers.OutputJSON(data, prettyPrint))
+}
+
+func getStackdriverResp(sdl StackDriverImpl, projectID string, metric string) (*monitoring.ListTimeSeriesResponse, error) {
+	startTime := time.Now().UTC().Add(time.Minute * -3)
```

Comment:
Do you know it will run every 3 minutes? Is there a way to configure it. This seems likely cause gaps in the data if anyone ever changes the run frequency.

Reply:
it runs every minute, datastore updates every minute with 3 minute old data, this is just to create a timestamp that reflects the timing of datastore